### PR TITLE
MLIR#283: update llvm (9/21/21)

### PR DESCRIPTION
**Pull request by sjw36**
RESUBMITTED: DO NOT SQUASH!!!

Changes in mlir (miopen):

- many similar std and dialect ops consolidated into memref dialect (alloc/dealloc/load/store/cast/reshape/subview/etc)
- moved math::<utility_funcs> to math_util namespace
- changed tosa-to-miopen to use memref.expand_shape
- minor refactor of RewritePatterns
- Much of BackendUtils is superceded by SerializeToHsaco
- updated tools for new SerializeToHsaco
- alot of test updates for memref dialect
- gpu-to-rocdl now requires gpu.func (vs func)

Changes in LLVM subtree:

- cleanup printers for miopen specific ops
- fixed SerializeToHsaco for target triple, chip, features
- Added back vector.tuple and vector.tuple_get impls
- Added mlir::registerToROCDLIRTranslation for -mlir-to-rocdlir

### CODE REVIEW

**Review by whchung — APPROVED** [review]

**Comment by whchung** [comment]
@sjw36 could you take a look at the failed CI targets? I have trouble getting into VPN now.

**Comment by whchung** [comment]
@sjw36 It seems all failed targets share the same symptom:

```
../lib/libMIOpen.so.1.0: undefined reference to `mlir::createGpuSerializeToHsacoPass(llvm::StringRef, llvm::StringRef, llvm::StringRef)'

clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
```

Could be a mismatch of pass interface, or a missing link target in one of the CMakeFiles?

**Comment by sjw36** [comment]
Yes, I think we will have to do this in 2 steps since this is in another repo (MIOpen).

**Comment by krzysz00** [comment]
> 
> 
> Yes, I think we will have to do this in 2 steps since this is in another repo (MIOpen).

I don't think the MIOpen repo uses our shared libraries?

**Comment by jerryyin** [comment]
@sjw36 In the future if this happens again I prefer this not to be merged. I lose the jenkins log once you merge the PR.

> Yes, I think we will have to do this in 2 steps since this is in another repo (MIOpen).

No from the look of it, it is populated from the static library, on our side. This is nothing we need to do in addition to MIOpen side so far I can tell.

File: mlir/tools/mlir-rocm-runner/mlir-rocm-runner.cpp [hunk 1]
```diff
@@ -51,6 +51,7 @@ void registerTestDialect(DialectRegistry &);
 } // namespace test
 
 static LogicalResult runMLIRPasses(ModuleOp m) {
+  m.getContext()->disableMultithreading();
```
Comment (jerryyin) [inline mlir/tools/mlir-rocm-runner/mlir-rocm-runner.cpp]: What was the problem that caused the need to disable multi threading? I thought we already put a lock on the linker?

Comment (sjw36) [reply mlir/tools/mlir-rocm-runner/mlir-rocm-runner.cpp]: Actually I don't think that is necessary. This was an attempt to fix the hang which turned out to be a node problem. I will revert this.

Comment (jerryyin) [reply mlir/tools/mlir-rocm-runner/mlir-rocm-runner.cpp]: Sounds good, hopefully no issue show up when we revert it.

**Pull request merged**